Drop usage of `.All` in Microsoft.Extensions

MichalStrehovsky opened the pull request:
The existing usage is a leftover from when we didn't have the binding source generation and the guidance was to suppress warning and preserve everything manually. We had a similar thing for S.T.Json in the past that we also deleted/dropped in favor of the source generation.

All the updated places are already marked RUC.

Cc @eerhardt - as you suggested on Teams

MichalStrehovsky committed:
Drop usage of `.All` in Microsoft.Extensions

The existing usage is a leftover from when we didn't have the binding source generation and the guidance was to suppress warning and preserve everything manually. We had a similar thing for S.T.Json in the past that we also deleted/dropped in favor of the source generation.

All the updated places are already marked RUC.

MichalStrehovsky committed:
Update src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

tarekgh commented:
@MichalStrehovsky can you look at the failures? https://github.com/dotnet/runtime/pull/117707/checks?check_run_id=46091806985.

Is there any app compat concerning this change? I assume not if we already did that before with S.T.Jeson.

MichalStrehovsky committed:
Fix ApiCompat errors

MichalStrehovsky commented:
> @MichalStrehovsky can you look at the failures? https://github.com/dotnet/runtime/pull/117707/checks?check_run_id=46091806985.

Fixed.

> Is there any app compat concerning this change? What kind of breaking would be? Also, could you please file a breaking change issue too?

Yup, that's why I added the breaking-change label. The break is that if someone is using this with trimming and without the source generator (and suppresses the trimming warnings that are generated for all of the methods this is updating), we'd previously keep members on the T (and presumably some of this still worked with trimming). Now switching to the source generator becomes a harder requirement. It can still be worked around by manual preservation and warning suppression. We don't recommend/support such use cases.

jeffhandley commented:
Breaking change document addressed by https://github.com/dotnet/runtime/pull/119865#issuecomment-3340686727
